zenoh/src/api/session.rs

```diff
@@ -1092,16 +1092,22 @@ impl SessionInner {
             } else {
                 primitives.send_close();
             }
+            // defer the cleanup of internal data structures by taking them out of the locked state
+            // this is needed because callbacks may contain entities which need to acquire the
+            // lock to be dropped, so callback must be dropped without the lock held
             let mut state = zwrite!(self.state);
-            state.queryables.clear();
-            state.subscribers.clear();
-            state.liveliness_subscribers.clear();
-            state.local_resources.clear();
-            state.remote_resources.clear();
+            let _queryables = std::mem::take(&mut state.queryables);
+            let _subscribers = std::mem::take(&mut state.subscribers);
+            let _liveliness_subscribers = std::mem::take(&mut state.liveliness_subscribers);
+            let _local_resources = std::mem::take(&mut state.local_resources);
+            let _remote_resources = std::mem::take(&mut state.remote_resources);
+            drop(state);
```

Comment:
Wouldn't be enough to drop the state at the end of the function?
Does the `unstable` feature need to drop first the state and then reacquiring it straight after?

Reply:
Fixed, it's simpler and clearer now.